test/test_tensor_spec.py

```diff
@@ -226,37 +230,150 @@ def test_mult_onehot(shape, ns):
         assert (ts.encode(np_r) == r).all()
 
 
+@pytest.mark.parametrize("is_complete", [True, False])
+@pytest.mark.parametrize("device", get_available_devices())
 @pytest.mark.parametrize("dtype", [torch.float32, torch.float16, torch.float64, None])
-@pytest.mark.parametrize(
-    "shape",
-    [
-        [],
-        torch.Size(
-            [
-                3,
-            ]
-        ),
-    ],
-)
-def test_composite(shape, dtype):
-    torch.manual_seed(0)
-    np.random.seed(0)
+class TestComposite:
+    @staticmethod
+    def _composite_spec(is_complete=True, device=None, dtype=None):
+        torch.manual_seed(0)
+        np.random.seed(0)
 
-    ts = CompositeSpec(
-        obs=NdBoundedTensorSpec(
-            torch.zeros(3, 32, 32), torch.ones(3, 32, 32), dtype=dtype
-        ),
-        act=NdUnboundedContinuousTensorSpec((7,), dtype=dtype),
-    )
-    if dtype is None:
-        dtype = torch.get_default_dtype()
+        return CompositeSpec(
+            obs=NdBoundedTensorSpec(
+                torch.zeros(3, 32, 32),
+                torch.ones(3, 32, 32),
+                dtype=dtype,
+                device=device,
+            ),
+            act=NdUnboundedContinuousTensorSpec((7,), dtype=dtype, device=device)
+            if is_complete
+            else None,
+        )
+
+    def test_getitem(self, is_complete, device, dtype):
+        ts = self._composite_spec(is_complete, device, dtype)
+        assert isinstance(ts["obs"], NdBoundedTensorSpec)
+        if is_complete:
+            assert isinstance(ts["act"], NdUnboundedContinuousTensorSpec)
+        else:
+            assert ts["act"] is None
+        with pytest.raises(KeyError):
+            _ = ts["UNK"]
+
+    def test_setitem_forbidden_keys(self, is_complete, device, dtype):
+        ts = self._composite_spec(is_complete, device, dtype)
+        for key in {"shape", "device", "dtype", "space"}:
+            with pytest.raises(AttributeError, match="cannot be set"):
+                ts[key] = 42
+
+    def test_setitem_matches_device(self, is_complete, device, dtype):
+        ts = self._composite_spec(is_complete, device, dtype)
+
+        bad_value, good_value = Mock(), Mock()
```

Comment:
Great idea of using Mock though I see two caveats:
At one point we might check that specs are or type TensorSpec, does mock works in that case? (Ie does isinstance(mock, type) always true for all types?)
Second we rely on pytest for our tests and I'm not sure if we want to mix unittest and pytest.